module/surface_sam_sid_vhf.c

```diff
@@ -0,0 +1,411 @@
+	case HID_FEATURE_REPORT:
+		switch (reqtype) {
+		case HID_REQ_GET_REPORT:
+			// The EC doesn't respond to GET FEATURE for these touchpad reports
+			// we immediately discard to avoid waiting for a timeout.
+			if (reportnum == 6 || reportnum == 7 || reportnum == 8 || reportnum == 9 || reportnum == 0x0b) {
+				hid_dbg(hid, "%s: skipping get feature report for 0x%02x\n", __func__, reportnum);
+				return 0;
+			}
+
+			cid = 0x02;
+			break;
+		case HID_REQ_SET_REPORT:
+			cid = 0x03;
+			break;
+		default:
+			hid_err(hid, "%s: unknown req type 0x%02x\n", __func__, rtype);
+			return 0;	// discard everything
+		}
+                break;
+	default:
+		hid_err(hid, "%s: unknown report type 0x%02x\n", __func__, reportnum);
+		return 0;		// discard everything
+	}
+
+        rqst.tc  = SAM_EVENT_SID_VHF_TC;
+        rqst.pri = SURFACE_SAM_PRIORITY_HIGH;
+        rqst.iid = 0x00; // windows tends to distinguish iids, but EC will take it
+        rqst.cid = cid;
+        rqst.snc = HID_REQ_GET_REPORT == reqtype ? 0x01 : 0x00;
+        rqst.cdl = HID_REQ_GET_REPORT == reqtype ? 0x01 : len;
+        rqst.pld = buf;
+
+        result.cap = len;
+        result.len = 0;
+        result.data = buf;
+
+	hid_dbg(hid, "%s: sending iid=%#04x cid=%#04x snc=%#04x\n", __func__, iid, cid, HID_REQ_GET_REPORT == reqtype);
+
+	status = surface_sam_ssh_rqst(&rqst, &result);
+	hid_dbg(hid, "%s: status %i\n", __func__, status);
+
+	if (status) {
+		return status;
+	}
+
+	if (result.len > 0) {
+		print_hex_dump_debug("response:", DUMP_PREFIX_OFFSET, 16, 1, result.data, result.len, false);
+	}
+
+	return result.len;
+}
+
+static struct hid_ll_driver sid_vhf_hid_ll_driver = {
+	.start         = sid_vhf_hid_start,
+	.stop          = sid_vhf_hid_stop,
+	.open          = sid_vhf_hid_open,
+	.close         = sid_vhf_hid_close,
+	.parse         = sid_vhf_hid_parse,
+	.raw_request   = sid_vhf_hid_raw_request,
+};
+
+
+static struct hid_device *sid_vhf_create_hid_device(struct platform_device *pdev, struct vhf_meta_meta_resp *meta)
+{
+	struct hid_device *hid;
+
+	hid = hid_allocate_device();
+	if (IS_ERR(hid)) {
+		return hid;
+	}
+
+	hid->dev.parent = &pdev->dev;
+
+	hid->bus     = BUS_VIRTUAL;
+	hid->vendor  = meta.vendor_id;
+	hid->product = meta.product_id;
+
+	hid->ll_driver = &sid_vhf_hid_ll_driver;
+
+	sprintf(hid->name, "%s", SID_VHF_INPUT_NAME);
+
+	return hid;
+}
+
+static int sid_vhf_event_handler(struct surface_sam_ssh_event *event, void *data)
+{
+	struct sid_vhf_evtctx *ctx = (struct sid_vhf_evtctx *)data;
+
+	if (event->tc == SAM_EVENT_SID_VHF_TC && (event->cid == 0x00 || event->cid == 0x03 || event->cid == 0x04)) {
+		return hid_input_report(ctx->hid, HID_INPUT_REPORT, event->pld, event->len, 1);
+	}
+
+	dev_warn(ctx->dev, "unsupported event (tc = %d, cid = %d)\n", event->tc, event->cid);
+	return 0;
+}
+
+static int surface_sam_sid_vhf_probe(struct platform_device *pdev)
+{
+	struct sid_vhf_drvdata *drvdata;
+	struct surface_sam_sid_vhf_meta_resp *meta;
```

Comment:
This should be `struct vhf_meta_meta_resp meta` according to the `vhf_get_metadata` signature. Also on stack.